core-blocks/image/edit.js

```diff
@@ -176,31 +193,53 @@ class ImageEdit extends Component {
 		return get( this.props.image, [ 'media_details', 'sizes' ], {} );
 	}
 
+	switchToEditing() {
+		const { url } = this.props.attributes;
+
+		this.props.setAttributes( {
+			url: undefined,
+			alt: undefined,
+			id: undefined,
+			caption: undefined,
+			src: url,
+		} );
+	}
+
 	render() {
-		const { attributes, setAttributes, isLargeViewport, isSelected, className, maxWidth, toggleSelection } = this.props;
-		const { url, alt, caption, align, id, href, width, height } = attributes;
+		const { attributes, setAttributes, isLargeViewport, isSelected, className, maxWidth, toggleSelection, hasUploadPermissions } = this.props;
+		const { url, alt, caption, align, id, href, width, height, src } = attributes;
+		const isLocal = ( url || src ) && ( url || src ).indexOf( window.location.origin ) === 0;
 
 		const controls = (
 			<BlockControls>
 				<BlockAlignmentToolbar
 					value={ align }
 					onChange={ this.updateAlignment }
 				/>
-
 				<Toolbar>
-					<MediaUpload
-						onSelect={ this.onSelectImage }
-						type="image"
-						value={ id }
-						render={ ( { open } ) => (
-							<IconButton
-								className="components-toolbar__control"
-								label={ __( 'Edit image' ) }
-								icon="edit"
-								onClick={ open }
-							/>
-						) }
-					/>
+					{ hasUploadPermissions && isLocal ? (
+						<MediaUpload
```

Comment:
I'm wondering it this logic should be enclosed inside `<MediaUpload />` component instead. There are 6 occurrences in the code where this component is used and we are covering only 2 occurrences in this PR. It seems like all of them should have a similar check. Moving those checks to the `MediaUpload` should make it easier to maintain.

Reply:
@imath Do you have an opinion on ^ this question?

Reply:
I'd say probably, but after this: https://github.com/WordPress/gutenberg/pull/4155/commits/200198f1a2e44b1da1c0f38e24396879826f3bb6 i'd also say that for the cover image the hasUploadPermissions check is before the `<toolbar>` tag and for image/video/audio it's after to fall back on a regular edit link. So i'm unsure about it 🤔

Reply:
There is a lot of code duplication at the moment, because you need to releat the same if statement and fetch the same data in all those places where you could donit only once in MediaUpload component. This is a general pattern we follow, data needs should be satisfied as close to the component that depends on it as possible.

As I understand isLocal check would have to stay as it is. 

Toolbar is not an issue in here, it will simply render nothing. There should be no visual difference, we can double check it. 

The only question we need to answer is if MediaUpload can ever be rendered if user doesn’t have a proper permission. If not, then I would be in favor of having this check more linked with MediaUpload. It doesn’t need to be the same file, we can create a wrapper component similar to the pattern used in the Document sidebar, example: https://github.com/WordPress/gutenberg/blob/master/edit-post/components/sidebar/post-author/index.js.

Reply:
Ok I will work on it.